File: flutter/lib/data/services/points_manager.dart
import 'package:cloud_firestore/cloud_firestore.dart';
import 'package:firebase_auth/firebase_auth.dart';
import 'package:flutter_riverpod/flutter_riverpod.dart';

/// ポイント管理サービス
class PointsManager {
  final FirebaseFirestore _db;
  final FirebaseAuth _auth;
  static const int pointsPerMessage = 10;

  PointsManager({
    FirebaseFirestore? firestore,
    FirebaseAuth? auth,
  })  : _db = firestore ?? FirebaseFirestore.instance,
        _auth = auth ?? FirebaseAuth.instance;

  /// ポイントを読み込み
  Future<int> loadPoints(String characterId) async {
    final userId = _auth.currentUser?.uid;
    if (userId == null || userId.isEmpty) {
      throw Exception('ユーザーが認証されていません');
    }

    if (characterId.isEmpty) {
      throw Exception('キャラクターIDが空です');
    }

    final doc = await _db
        .collection('users')
        .doc(userId)
        .collection('characters')
        .doc(characterId)
        .collection('details')
        .doc('current')
        .get();

    if (doc.exists && doc.data() != null) {
      return doc.data()!['points'] as int? ?? 0;
    }
    return 0;
  }

  /// ポイントを付与（チャット送信時）
  Future<int> addPoints(String characterId, {int? currentPoints}) async {
    final userId = _auth.currentUser?.uid;
    if (userId == null || userId.isEmpty) {
      throw Exception('ユーザーが認証されていません');
    }

    if (characterId.isEmpty) {
      throw Exception('キャラクターIDが空です');
    }

    final points = currentPoints ?? await loadPoints(characterId);
    final newPoints = points + pointsPerMessage;

    final detailsRef = _db
        .collection('users')
        .doc(userId)
        .collection('characters')
        .doc(characterId)
        .collection('details')
        .doc('current');

    final doc = await detailsRef.get();

    if (doc.exists) {
      await detailsRef.update({
        'points': newPoints,
        'updated_at': FieldValue.serverTimestamp(),
      });
    } else {
      await detailsRef.set({
        'points': newPoints,
        'created_at': FieldValue.serverTimestamp(),
        'updated_at': FieldValue.serverTimestamp(),
      }, SetOptions(merge: true));
    }

    return newPoints;
  }

  /// ポイントを消費
  Future<int> consumePoints(
    String characterId,
    int amount, {
    int? currentPoints,
  }) async {
    final userId = _auth.currentUser?.uid;
    if (userId == null || userId.isEmpty) {
      throw Exception('ユーザーが認証されていません');
    }

    if (characterId.isEmpty) {
      throw Exception('キャラクターIDが空です');
    }

    final points = currentPoints ?? await loadPoints(characterId);
    if (points < amount) {
      throw Exception('ポイントが不足しています');
    }

    final newPoints = points - amount;

    final detailsRef = _db
        .collection('users')
        .doc(userId)
        .collection('characters')
        .doc(characterId)
        .collection('details')
        .doc('current');

    await detailsRef.update({
      'points': newPoints,
      'updated_at': FieldValue.serverTimestamp(),
    });

    return newPoints;
  }

  /// ポイントを監視するストリーム
  Stream<int> watchPoints(String characterId) {
    final userId = _auth.currentUser?.uid;
    if (userId == null || userId.isEmpty) {
      return Stream.value(0);
    }

    if (characterId.isEmpty) {
      return Stream.value(0);
    }

    return _db
        .collection('users')
        .doc(userId)
        .collection('characters')
        .doc(characterId)
        .collection('details')
        .doc('current')
        .snapshots()
        .map((doc) {
      if (doc.exists && doc.data() != null) {
        return doc.data()!['points'] as int? ?? 0;
      }
      return 0;
    });
  }
}

/// プロバイダー
final pointsManagerProvider = Provider<PointsManager>((ref) {
  return PointsManager();
});

/// キャラクターのポイントを監視するプロバイダー
final characterPointsProvider =
    StreamProvider.family<int, String>((ref, characterId) {
  final manager = ref.watch(pointsManagerProvider);
  return manager.watchPoints(characterId);
});

File: ios/Character/Services/PointsManager.swift
import Foundation
import FirebaseFirestore
import FirebaseAuth

class PointsManager: ObservableObject {
    @Published var currentPoints: Int = 0
    
    private let db = Firestore.firestore()
    private let pointsPerMessage = 10 // 1メッセージあたり10ポイント
    
    // ポイント読み込み
    func loadPoints(for characterId: String) {
        guard let currentUserId = Auth.auth().currentUser?.uid, !currentUserId.isEmpty else {
            Logger.error("User not authenticated for points loading", category: Logger.authentication)
            return
        }

        guard !characterId.isEmpty else {
            Logger.error("CharacterId cannot be empty for points loading", category: Logger.general)
            return
        }
        
        db.collection("users").document(currentUserId)
            .collection("characters").document(characterId)
            .collection("details").document("current").getDocument { [weak self] document, error in
                if let error = error {
                    return
                }
                
                if let document = document, document.exists {
                    let points = document.data()?["points"] as? Int ?? 0
                    DispatchQueue.main.async {
                        self?.currentPoints = points
                    }
                }
            }
    }
    
    // ポイント付与（チャット送信時）
    func addPoints(for characterId: String, completion: @escaping (Bool) -> Void = { _ in }) {
        guard let currentUserId = Auth.auth().currentUser?.uid, !currentUserId.isEmpty else {
            Logger.error("User not authenticated for points adding", category: Logger.authentication)
            completion(false)
            return
        }

        guard !characterId.isEmpty else {
            Logger.error("CharacterId cannot be empty for points adding", category: Logger.general)
            completion(false)
            return
        }
        
        let newPoints = currentPoints + pointsPerMessage
        
        let detailsRef = db.collection("users").document(currentUserId)
            .collection("characters").document(characterId)
            .collection("details").document("current")
        
        // ドキュメントの存在確認とデータ更新
        detailsRef.getDocument { document, error in
            if let error = error {
                completion(false)
                return
            }
            
            if document?.exists == true {
                // ドキュメントが存在する場合は更新
                detailsRef.updateData([
                    "points": newPoints,
                    "updated_at": Timestamp()
                ]) { [weak self] error in
                    if let error = error {
                        completion(false)
                    } else {
                        DispatchQueue.main.async {
                            self?.currentPoints = newPoints
                        }
                        completion(true)
                    }
                }
            } else {
                // ドキュメントが存在しない場合は作成
                detailsRef.setData([
                    "points": newPoints,
                    "created_at": Timestamp(),
                    "updated_at": Timestamp()
                ], merge: true) { [weak self] error in
                    if let error = error {
                        completion(false)
                    } else {
                        DispatchQueue.main.async {
                            self?.currentPoints = newPoints
                        }
                        completion(true)
                    }
                }
            }
        }
    }
    
    // ポイント消費（将来の機能拡張用）
    func consumePoints(for characterId: String, amount: Int, completion: @escaping (Bool) -> Void = { _ in }) {
        guard let currentUserId = Auth.auth().currentUser?.uid, !currentUserId.isEmpty else {
            Logger.error("User not authenticated for points consumption", category: Logger.authentication)
            completion(false)
            return
        }

        guard !characterId.isEmpty else {
            Logger.error("CharacterId cannot be empty for points consumption", category: Logger.general)
            completion(false)
            return
        }
        
        guard currentPoints >= amount else {
            completion(false)
            return
        }
        
        let newPoints = currentPoints - amount
        
        let detailsRef = db.collection("users").document(currentUserId)
            .collection("characters").document(characterId)
            .collection("details").document("current")
        
        // ドキュメントの存在確認とデータ更新
        detailsRef.getDocument { document, error in
            if let error = error {
                completion(false)
                return
            }
            
            if document?.exists == true {
                // ドキュメントが存在する場合は更新
                detailsRef.updateData([
                    "points": newPoints,
                    "updated_at": Timestamp()
                ]) { [weak self] error in
                    if let error = error {
                        completion(false)
                    } else {
                        DispatchQueue.main.async {
                            self?.currentPoints = newPoints
                        }
                        completion(true)
                    }
                }
            } else {
                completion(false)
            }
        }
    }
    
    // ポイント監視開始
    func startPointsListener(for characterId: String) {
        guard let currentUserId = Auth.auth().currentUser?.uid, !currentUserId.isEmpty else {
            Logger.error("User not authenticated for points monitoring", category: Logger.authentication)
            return
        }

        guard !characterId.isEmpty else {
            Logger.error("CharacterId cannot be empty for points monitoring", category: Logger.general)
            return
        }
        
        db.collection("users").document(currentUserId)
            .collection("characters").document(characterId)
            .collection("details").document("current")
            .addSnapshotListener { [weak self] documentSnapshot, error in
                if let error = error {
                    return
                }
                
                if let document = documentSnapshot, document.exists {
                    let points = document.data()?["points"] as? Int ?? 0
                    DispatchQueue.main.async {
                        self?.currentPoints = points
                    }
                }
            }
    }
}

// NotificationCenter用の拡張
extension Notification.Name {
    static let pointsUpdated = Notification.Name("pointsUpdated")
}